pkg/monitor/datapath_drop.go

```diff
@@ -34,38 +37,53 @@ type DropNotify struct {
 	Hash     uint32
 	OrigLen  uint32
 	CapLen   uint32
-	SrcLabel uint32
-	DstLabel uint32
+	SrcLabel identity.NumericIdentity
+	DstLabel identity.NumericIdentity
 	DstID    uint32
 	Unused   uint32
 	// data
 }
 
+// dumpIdentity dumps the source and destination identities in numeric or
+// human-readable format.
+func (n *DropNotify) dumpIdentity(buf *bufio.Writer, numeric bool) {
+	if numeric {
+		fmt.Fprintf(buf, "identity %d->%d", n.SrcLabel, n.DstLabel)
+	} else {
+		fmt.Fprintf(buf, "identity %s->%s", n.SrcLabel, n.DstLabel)
+	}
+}
+
 // DumpInfo prints a summary of the drop messages.
-func (n *DropNotify) DumpInfo(data []byte) {
-	fmt.Printf("xx drop (%s) flow %#x to endpoint %d, identity %d->%d: %s\n",
-		api.DropReason(n.SubType), n.Hash, n.DstID, n.SrcLabel, n.DstLabel,
-		GetConnectionSummary(data[DropNotifyLen:]))
+func (n *DropNotify) DumpInfo(data []byte, numeric bool) {
+	buf := bufio.NewWriter(os.Stdout)
+	fmt.Fprintf(buf, "xx drop (%s) flow %#x to endpoint %d, ",
+		api.DropReason(n.SubType), n.Hash, n.DstID)
+	n.dumpIdentity(buf, numeric)
+	fmt.Fprintf(buf, ": %s\n", GetConnectionSummary(data[DropNotifyLen:]))
+	buf.Flush()
```

Comment:
This code hardwires `os.Stdout` as a destination into the function. The code will be more flexible and testable if instead an `io.Writer` is passed in and the caller choses the destination.

Reply:
I agree on the general idea, but I'm not sure it's worth it in this case. 1) It's not a regression introduced by the new code changes, and more importantly, 2) it's unlikely we'll ever unit test those functions because they are very simple. Passing the destination as an argument of the function will make the caller's code a bit more complex and will add yet another argument to those functions.